topics/go/algorithms/data/heap/heap.go

```diff
@@ -0,0 +1,158 @@
+// Package heap implements a heap interface, and add the ability to store values
+// of type Data in the minimum Heap.
+package heap
+
+import (
+	"errors"
+)
+
+// Data represents what is being stored on the heap.
+type Data struct {
+	Value int
+	Index int
+}
+
+// Heap represents a list of data.
+type Heap struct {
+	capacity int
+	data []Data
+}
+
+// New returns a new Heap with in which we will store our data.
+func New(cap int) (*Heap, error) {
+	if cap <= 0 {
+		return nil, errors.New("invalid capacity")
+	}
+	h := Heap{
+		capacity: cap,
+	}
+
+	return &h, nil
+}
+
+// Store is storing the element of type Data in the heap.
+func (h *Heap) Store(d Data) error {
+	if len(h.data) > h.capacity {
+		return errors.New("out of heap capacity")
+	}
+	// The element which we will be store in the heap will be added closely to
```

Comment:
My idea is we return the data using pointer semantics, then they need to release it using the point. But we maintain the contigous block of memory.

We add a New or Malloc function that returns an address to a Data value.
We add a Delete or Free function that takes the address to release it.